20220714-lookup-api.md

```diff
@@ -0,0 +1,457 @@
+
+* [Search with Permissions](https://openfga.dev/docs/interacting/search-with-permissions) - Given a particular search filter (and optional sort order), what objects can the user acces?
+* [Authorization Model](https://openfga.dev/docs/concepts#what-is-an-authorization-model)
+* [Relationship Tuples](https://openfga.dev/docs/concepts#what-is-a-relationship-tuple)
+* [gRPC Server Streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc)
+* [gRPC Unary RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc)
+
+# Motivation
+[motivation]: #motivation
+
+- **Why should we do this?**
+  It simplifies client integrations with OpenFGA for Search with Permissions and it adddress an ask from the community for an endpoint to list the objects a user has permissions for.
+
+- **What use cases does it support?**
+  Search with Permissions where the number of objects of a particular type the user could have access to is low (~ thousands), and the percentage of the total objects that the user can have access to is also low.
+
+- **What is the expected outcome?**
+  OpenFGA clients use the Lookup API to return a list of object ids and then the client uses the list to filter and order their object list in their applications (pagination on the filtered list is the clients responsibility). The expectation is that the client consumes the whole Lookup response before filtering client side, and does not try to partially filter based on a subset of the response from Lookup. Hence, it is not the goal of Lookup to be able to continue the Lookup process from a prior invocation.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The Lookup API will provide clients with the ability to query the object ids that a user has a particular relationship with. The list of object ids the Lookup API returns can then be used by a client to filter an object collection to only the objects that a user has a particular relationship with by taking the intersection of it with any client provided filters/ordering of the underlying dataset. This is described in more detail with [Search with Permission (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search) document.
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+```
+document:doc1, viewer, bob
```

Comment:
Can we update this to be clearer?

- user=bob, relation=editor, object=document:doc2
and so on

Reply:
I changed these to a markdown table. How does that look?

Reply:
A markdown table is much better. Though I would still use the order of `user, relation, object` to be more consistent with how we use them pretty much across all docs. Also generally, people found it much easier to read when put this way.